### PR TITLE
NAS-100851 / 12 / pool.expand

### PR DESCRIPTION
**Pull request by themylogin**

### CODE REVIEW

File: src/middlewared/middlewared/plugins/pool.py [hunk 1]
```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
...
+    )
+    @job(lock="pool_expand")
+    async def expand(self, job, id, options):
+        """
+        Expand pool to fit all available disk space.
+        """
+
+        pool = await self._get_instance(id)
+
+        if pool["encrypt"]:
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Whole file seems to be using single quotes, lets keep it that way please : P

File: src/middlewared/middlewared/plugins/pool.py [hunk 2]
```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
...
+                            "geli", "resize", "-s", str(mediasize), vdev["device"]
+                        )
+                        rollback_cmd = (
+                            "gpart", "resize", "-i", str(partition_number), "-s", str(mediasize), vdev["disk"]
+                        )
+
+                        logger.warning("It will be obligatory to notify GELI that the provider has been resized: %r",
+                                       join_commandline(geli_resize_cmd))
+                        logger.warning("Or to resize provider back: %r",
+                                       join_commandline(rollback_cmd))
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Whats the intent of logging about this?

File: src/middlewared/middlewared/plugins/pool.py [hunk 3]
```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
...
+
+                        geli_resize.append((geli_resize_cmd, rollback_cmd))
+            finally:
+                if geli_resize:
+                    failed_rollback = []
+
+                    try:
+                        lock_job = await self.middleware.call("pool.lock", pool["id"], options["geli"]["passphrase"])
+                        await lock_job.wait()
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: wait() does not raise any exception from what I recall

File: src/middlewared/middlewared/plugins/pool.py [hunk 4]
```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
...
+                            def copy():
+                                with open(pool["encryptkey_path"], "rb") as f:
+                                    shutil.copyfileobj(f, unlock_job.pipes.input.w)
+
+                            try:
+                                await self.middleware.run_in_thread(copy)
+                            finally:
+                                await self.middleware.run_in_thread(unlock_job.pipes.input.w.close)
+
+                        await unlock_job.wait()
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Check for error if unlock failed?

File: src/middlewared/middlewared/plugins/pool.py [hunk 5]
```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
...
+                                                 join_commandline(geli_resize_cmd),
+                                                 geli_resize_error.stderr.strip())
+                                    if not await self.__run_rollback_cmd(rollback_cmd):
+                                        failed_rollback.append(rollback_cmd)
+
+                        if failed_rollback:
+                            raise CallError("Resizing partitions of your encrypted pool failed and rolling back "
+                                            "changes failed too. You'll need to run the following commands manually:\n"
+                                            "%s",
+                                            "\n".join(map(join_commandline, failed_rollback)))
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: Dont we already show the command in __run_rollback_cmd ?

File: src/middlewared/middlewared/plugins/pool.py [hunk 6]
```diff
@@ -2177,6 +2185,157 @@ def import_disk_msdosfs_locales(self):
...
+
+                    partition_number = RE_DISKPART.match(vdev["device"])
+                    if partition_number is None:
+                        logger.debug("Could not parse partition number from %r", vdev["device"])
+                        continue
+
+                    assert partition_number.group(1) == vdev["disk"]
+                    partition_number = int(partition_number.group(2)[1:])
+
+                    mediasize = bsd.geom.geom_by_name("LABEL", vdev["device"]).provider.mediasize
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/pool.py]: you need to use geom.scan() to make sure its up-to-date.

**Comment by william-gr** [comment]
I only saw later on that you had removed the request for review and added WIP ;)

File: src/middlewared/middlewared/plugins/pool.py (hunk 2 above)
Comment (themylogin) [reply src/middlewared/middlewared/plugins/pool.py]: We need to run `geli resize` with old size of partition or the data is lost. That's why it's logged ASAP (in very rare case when middleware crashes).

File: src/middlewared/middlewared/plugins/pool.py (hunk 5 above)
Comment (themylogin) [reply src/middlewared/middlewared/plugins/pool.py]: `__run_rollback_cmd` logs it to middleware log, here we output it to user.

**Pull request merged**